svc/GridAutosizeService.js

```diff
@@ -0,0 +1,143 @@
+/*
+ * This file belongs to Hoist, an application development toolkit
+ * developed by Extremely Heavy Industries (www.xh.io | info@xh.io)
+ *
+ * Copyright © 2020 Extremely Heavy Industries Inc.
+ */
+
+import {HoistService} from '@xh/hoist/core';
+import {throwIf, stripTags} from '@xh/hoist/utils/js';
+import {isFunction, isNil, isFinite, sortBy, reduce} from 'lodash';
+
+@HoistService
+export class GridAutosizeService {
+
+    _canvasContext;
+    _cellEl;
+
+    autoSizeColumns({gridModel, colIds = []}) {
+        throwIf(!gridModel, 'GridModel required for autosize');
+        if (!colIds.length) return;
+
+        const ret = [];
+        for (let i = 0; i < colIds.length; i++) {
+            const colId = colIds[i],
+                width = this.autoSizeColumn({colId, gridModel});
+
+            if (isFinite(width)) ret.push({colId, width});
+        }
+        return ret;
+    }
+
+    //------------------
+    // Implementation
+    //------------------
+    autoSizeColumn({colId, gridModel}) {
+        try {
+            const {store} = gridModel,
+                records = [...store.records, store.summaryRecord],
+                column = gridModel.findColumn(gridModel.columns, colId),
+                {autoSizeOptions, field, getValueFn, renderer, elementRenderer} = column,
+                {enabled, sampleCount, buffer, minWidth, maxWidth} = autoSizeOptions,
+                useRenderer = isFunction(renderer);
+
+            if (!enabled) return;
+
+            // Columns with element renderers are not supported
+            if (elementRenderer) return;
+
+            // 1) Get unique values
+            const values = new Set();
+            records.forEach(record => {
+                if (!record) return;
+                const rawValue = getValueFn({record, field, column, gridModel}),
+                    value = useRenderer ? renderer(rawValue, {record, column, gridModel}) : rawValue;
+                values.add(value);
+            });
+
+            // 2) Use a canvas to estimate and sort by the pixel width of the string value.
+            // Strip html tags but include parentheses / units etc. for renderers that may return html,
+            const sortedValues = sortBy(Array.from(values), value => {
+                return isNil(value) ? 0 : this.getStringWidth(stripTags(value.toString()));
+            });
+
+            // 3) Extract the sample set of longest values for rendering and sizing
+            const longestValues = sortedValues.slice(Math.max(sortedValues.length - sampleCount, 0));
```

Comment:
Tom and I discussed this, but still wondering if specifying sampleCount as a percentage might be more appropriate?  Or (not to overdesign this) but also allow specifying either a number like '1%' or 10. (in the latter case, would be sure to want a minimum)